### ISSUE TITLE
実装: 競り価値を役職確率・両陣営獲得枚数・契約成功/marginへ分解する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#378]
Related: #202, #373, #375, #377
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #375 では、競りstate/actionから最終役職と `candidateTeamPointCards` を予測する role-decomposed model を実装した。

Issue #377 では、1k/2k/5k/10k/20k source states の fixed-validation learning curve を取得したが、sample数を増やしても validation はほぼ頭打ちだった。

主な結果（1k -> 20k）:

- MAE: 3.233 -> 3.239
- RMSE: 4.066 -> 4.053
- Pearson: 0.319 -> 0.342
- role accuracy: 93.97% -> 94.97%
- best-action hit: 35.60% -> 35.10%
- pairwise ranking: 54.82% -> 55.36%
- strongest-suit match: 34.80% -> 32.00%

したがって、同一setupでsample数だけ増やす優先度は低い。

また `candidateTeamPointCards` は、candidateがNapoleon/Adjutant/Citizenのどの役職かによって意味が変わる値を1本にまとめているため、action価値を学ぶ教師として情報を混ぜすぎている可能性がある。

## 目的

競りstate `s` とaction `a` に対して、少なくとも以下を別々に予測するmulti-head modelを実装・評価する。

1. `P(finalRole | s,a)`
2. `E(napoleonSidePointCards | s,a, role/outcome context)`
3. `E(coalitionSidePointCards | s,a, role/outcome context)`
4. `P(contractSuccess | s,a, contract context)`
5. `E(contractMargin | s,a, contract context)`

これらを分離して学習することで、単一 `candidateTeamPointCards` より state/action に応じたsuit/target signalを抽出できるか確認する。

このIssueでは、最終runtime bidding scoreを確定・接続することより、各headが十分学習可能かをまず検証する。

## 重要: future leakage禁止

以下はすべて教師ラベルであり、推論時入力へ混入してはいけない。

- finalRole
- finalNapoleon
- finalAdjutant
- napoleonSidePointCards
- coalitionSidePointCards
- contractSuccess
- contractMargin
- finalDeclaredTarget / finalDeclaredSuit（action時点で未知な将来情報である場合）

入力は現行 bidding compact278 + forced/current action semantics の範囲に限定する。

## Dataset

Issue #375 schema v2を再利用する。

既にraw sampleに以下があることを確認する。

- finalRole
- candidateFinalTeam
- napoleonSidePointCards
- coalitionSidePointCards
- candidateTeamPointCards
- finalDeclaredTarget
- finalDeclaredSuit
- contractSuccess
- contractMargin

不足があれば最小限schemaをversion-upする。

## 対戦環境

Issue #375/#377と同じ条件を維持する。

Frozen bidding opponents:

- RuleBased 50%
- Conservative 50%
- frozen seatごとに独立
- seeded deterministic

Downstream:

- playing: `ppo-separated-v1000`
- adjutant/exchange: current frozen artifacts

まず既存 #377 full poolを再利用できるか確認する。

`/tmp/napoleon-issue377-qdataset-22000x1`

利用可能なら、新規rolloutを作らずこのpoolを使う。

## Model

Issue #375のrole-decomposed networkを拡張する。

推奨概念構造:

```text
compact278
   -> shared MLP
      -> 29 x R role logits
      -> 29 x ? Napoleon-side point-card prediction
      -> 29 x ? Coalition-side point-card prediction
      -> 29 x ? contract-success logits/probability
      -> 29 x ? contract-margin regression
```

role-conditionedにするべきheadと、action-conditionedのみでよいheadを実データ・意味論に基づいて整理する。

過度に複雑化しないこと。

## Head 1: final role

既存 #375 の role headを維持・再利用する。

評価:

- accuracy
- macro F1
- cross entropy
- per-role precision/recall
- confusion matrix
- majority baseline
- action-index baseline

## Head 2: Napoleon-side point cards

最終的にNapoleon側が獲得したポイント札枚数を予測する。

NoContract/AllPassでは通常回帰lossからmaskする。

最低限:

- MAE
- RMSE
- Pearson
- role別
- suit別
- target別

を出す。

## Head 3: Coalition-side point cards

連合軍側獲得ポイント札枚数を別headで予測する。

Napoleon-sideと相補関係がある場合、そのルール上の関係を確認する。

完全に `total - napoleonSide` で決まるなら、独立headを持つ価値があるかを検討し、必要ならNapoleon-sideのみ予測 + deterministic deriveでもよい。

その判断理由をIssueへ残す。

## Head 4: contract success

最終契約が存在するsampleについて、Napoleon側が宣言枚数を達成したかを分類する。

- binary classification
- BCE / equivalent
- NoContractはmask

評価:

- accuracy
- precision / recall / F1
- ROC-AUCまたはPR-AUC（実装容易なら）
- calibration
- suit別/target別success prediction

単純なtarget/action-index success-rate baselineとも比較する。

## Head 5: contract margin

契約が存在するsampleで:

```text
contractMargin = napoleonSidePointCards - finalDeclaredTarget
```

を予測する。

評価:

- MAE
- RMSE
- Pearson
- sign accuracy（成功/失敗方向）
- suit別
- target別

NoContractはmaskする。

## 学習loss

multi-task learningとする場合、各lossのscaleが違うため重み付けを明示する。

最低限:

- role classification loss
- Napoleon-side point-card regression loss
- coalition-side loss（採用する場合）
- contract success loss
- contract margin loss

を個別にreportする。

初版のweightは1.0固定でもよいが、loss scale差で一headが支配していないか確認する。

必要ならtarget standardization / loss weight調整をoption化する。

## Train / validation

#377 と比較可能な fixed-validation 条件を優先する。

可能なら #377 のvalidation stateKey集合:

`9ff43c57867712f5352467b3d56e397046b02c3c11b22a579ab0298a9777d57a`

を再利用する。

trainingはまず 20k poolを使用する。

ただし #377でfixed epochsによるtraining metric改善とvalidation頭打ちが見えたため、今回は過学習を避ける目的で以下も検討する。

- validation based early stopping
- weight decay
- dropout（必要なら）
- fixed optimizer step comparison

最初から大規模hyperparameter sweepはしない。

## Baseline

各headごとに単純baselineを作る。

### Role
- majority
- action-index role frequency

### Point cards
- global mean
- role mean
- action-index mean
- role x action mean

### Contract success
- global success rate
- target/suit/action-index success rate

### Margin
- global mean
- target mean
- action-index mean

NNがstate情報を使えているかをbaselineとの差で判断する。

## Teacher signal diagnostics

学習前に、同一stateのteacher-covered actionsについて以下を再確認する。

- Napoleon-side point-card best-worst gap
- coalition-side point-card gap
- contract success action差
- contract margin action差
- suit間exact tie率
- target間exact tie率
- strongest suit teacher-best率

`candidateTeamPointCards` よりどのteacherがsuit/target差を最も明確に持つかを比較する。

## Action score候補のoffline検証

各headが学習できた場合のみ、runtime接続せずofflineでaction score候補を試す。

例えば:

```text
score(s,a)
 = f(
     P(role|s,a),
     E(napoleonSidePointCards|s,a),
     P(contractSuccess|s,a),
     E(contractMargin|s,a)
   )
```

ただし恣意的なweight tuningで結果を作らない。

まずは以下のような単純候補を比較する。

- expected candidate-team cards相当を再構成
- Napoleonになる確率 x Napoleon-side cards
- success probability中心
- margin中心

PASS/Adjutant/Citizenをどう評価するかはrole probabilityを使って明示する。

このIssueで最終scoreを決定できなくてもよい。

## Evaluation

teacher-covered action subset内で:

- best-action hit
- top-3 hit
- pairwise ranking accuracy
- PASS vs best-bid
- predicted best suit distribution
- predicted best target distribution
- strongest suit match

を #375/#377 と比較する。

ただし「何をbestとするか」は各teacher/scoreごとに明示する。

## Success criteria

この方向が有望と判断する条件:

- Napoleon-side point-card predictionが candidateTeamPointCards よりbaseline差を広げる
- contract success / marginに明確なstate-conditioned signalがある
- PearsonやMAEだけでなくaction rankingが改善する
- suit/target rankingが55%前後のpairwise ceilingを超える
- strongest-suit関係がrandom付近から改善する
- 20k datasetでもvalidation improvementが確認できる

## Failure / next step

以下ならteacher分解だけでは不足と判断する。

- 各head単独でもbaselineをほぼ超えない
- action rankingが改善しない
- suit/target signalが依然弱い

その場合は次に:

- compact278 observation内容
- hand representation
- suit-relative / symmetry-aware architecture
- model architecture
- contested auction state sampling

を調査する。

## Tests

最低限:

1. schema loader / backward compatibility
2. future leakageなし
3. role head shape/loss
4. Napoleon-side cards head shape/loss
5. coalition-side teacher/derive fixture
6. NoContract regression mask
7. contract success label/loss
8. contract margin label/loss
9. multi-loss aggregation
10. stateKey split leakageなし
11. baseline metrics fixture
12. ranking diagnostics fixture
13. deterministic CPU smoke
14. checkpoint save/load
15. ONNX exportする場合はPyTorch parity

## Git / PR運用

恒久コード変更が必要なので、最新 `feature/issue-202-non-playing-ai` からIssue専用branchを作成する。

PR baseは必ず `feature/issue-202-non-playing-ai`。
main向けPRは作らない。

実装・tests・small/full experiment後、Codex review最大4回（review後修正最大3回）。blocking/actionable指摘がなくなればReadyにして親branchへmergeする。

mainにはmergeしない。

## 完了時に必ず報告

- dataset identity / state/sample数
- fixed validation identity
- model architecture
- each head definition
- each head loss/weight
- role metrics + baselines
- Napoleon-side cards MAE/RMSE/Pearson + baselines
- coalition-side prediction/derive結果
- contract success metrics + baselines
- contract margin MAE/RMSE/Pearson + baselines
- suit/target別metrics
- teacher exact tie/gap diagnostics
- offline action-score候補ごとのranking
- #375/#377との比較
- strongest suit match
- この分解が有望か
- runtime score設計へ進めるか
- それともobservation/architecture調査へ進むべきか
- PR / merge commit / Codex review回数
- mainへmergeしていないこと
